Add assertions to ensure memcpySpan and memsetSpan are used on trivially copyable spans

mcatanzaro opened the pull request:
#### 3ae0cd09762de6f9fd715033fc108bdf5fc1d3d0
<pre>
Add assertions to ensure memcpySpan and memsetSpan are used on trivially copyable spans
<a href="https://bugs.webkit.org/show_bug.cgi?id=264619">https://bugs.webkit.org/show_bug.cgi?id=264619</a>

Reviewed by Darin Adler.

Bad things are likely to happen if used on objects with custom
constructors or destructors or assignment operators, because they won&apos;t
be executed. Add safety checks for this.

* Source/WTF/wtf/Algorithms.h:
(WTF::memcpySpan):
(WTF::memsetSpan):

Canonical link: <a href="https://commits.webkit.org/270796@main">https://commits.webkit.org/270796@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/142fa219e54a6cb5f918647608af9a8287aa9286

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25785 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27069 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27883 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23610 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6153 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1828 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23722 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26035 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3298 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22223 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/2931 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2920 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23178 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28463 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/22440 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23532 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23550 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29248 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25008 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2939 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1172 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27117 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4326 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/32448 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6350 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3392 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7052 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3254 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
Nice idea for `memcpySpan`. Not sure for `memsetSpan`. There may be objects that can be efficiently initialized to 0 even if they aren’t trivially copyable. But presumably does no harm. We might want additional assertions, like trivially destructible?

mcatanzaro committed:
Add assertions to ensure memcpySpan and memsetSpan are used on trivially copyable spans
https://bugs.webkit.org/show_bug.cgi?id=264619

Reviewed by Darin Adler.

Bad things are likely to happen if used on objects with custom
constructors or destructors or assignment operators, because they won't
be executed. Add safety checks for this.

* Source/WTF/wtf/Algorithms.h:
(WTF::memcpySpan):
(WTF::memsetSpan):

Canonical link: https://commits.webkit.org/270796@main
